go/cmd/ct-fetch/ct-fetch.go

```diff
@@ -79,6 +82,167 @@ type CtLogEntry struct {
 	LogURL   string
 }
 
+type CtLogSubtree struct {
+	Root  []byte
+	First uint64
+	Last  uint64
+}
+
+func (r *CtLogSubtree) Size() uint64 {
+	return r.Last - r.First + 1
+}
+
+func (r *CtLogSubtree) Midpoint() uint64 {
+	size := r.Size()
+	prevPow2 := uint64(0)
+	if size > 1 {
+		prevPow2 = 1 << (bits.Len64(size-1) - 1)
+	}
+	return r.First + prevPow2
+}
+
+func rfc6962LeafHash(leaf []byte) []byte {
+	// Specified in section 2.1 of RFC 6962
+	h := crypto.SHA256.New()
+	h.Write([]byte{0})
+	h.Write(leaf)
+	return h.Sum(nil)
+}
+
+func rfc6962PairHash(left, right []byte) []byte {
+	// Specified in section 2.1 of RFC 6962
+	h := crypto.SHA256.New()
+	h.Write([]byte{1})
+	h.Write(left)
+	h.Write(right)
+	return h.Sum(nil)
+}
+
+type CtLogSubtreeVerifier struct {
+	Subtree     CtLogSubtree
+	hashStack   [][]byte // Scratch space for computing tree hash
+	numConsumed uint64   // Number of leaves hashed into hashStack
+}
+
+func (v *CtLogSubtreeVerifier) Consume(leaf []byte) {
+	if v.numConsumed == 0 {
+		// The hash stack might need to store a full sibling path
+		v.hashStack = make([][]byte, 0, bits.Len64(v.Subtree.Size())+2)
+	}
+	// Push the new leaf hash, H(0 || leaf), onto the stack
+	v.hashStack = append(v.hashStack, rfc6962LeafHash(leaf))
+	v.numConsumed += 1
+
+	// Iteratively pop a pair of siblings off the stack and replace them
+	// by H(1 || Left || Right). The number of times we do this is equal
+	// to the height of the largest complete subtree that contains the leaf
+	// that we just consumed.
+	iter := bits.TrailingZeros64(v.numConsumed)
+	for iter > 0 {
+		n := len(v.hashStack) - 1
+		L := v.hashStack[n-1]
+		R := v.hashStack[n]
+		v.hashStack = v.hashStack[:n-1]
+		v.hashStack = append(v.hashStack, rfc6962PairHash(L, R))
+		iter -= 1
+	}
+}
+
+func (v *CtLogSubtreeVerifier) CheckClaim() error {
+	if len(v.Subtree.Root) != crypto.SHA256.Size() {
+		return fmt.Errorf("CtLogSubtreeVerifier: Claim has the wrong length.")
+	}
+	if v.numConsumed != v.Subtree.Size() {
+		return fmt.Errorf("CtLogSubtreeVerifier: Consumed %d leaves but needed %d.", v.numConsumed, v.Subtree.Size())
+	}
+	// Finish computing the Merkle tree head.
+	for len(v.hashStack) > 1 {
```

Comment:
Yeah, I feel like `Consume` should be responsible for this.

Reply:
I agree. I tacked the change onto Part 4: 29c1c3cc13fd7dae7ad35cacfb8e1ef9a81feb16

Reply:
👍